### PR TITLE
Week 4 Authentication

### PR DESCRIPTION
**Pull request by BenLin00**
doGet seems to automatically run (?) whenever doPost is called. I'm not sure why.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -45,8 +45,24 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     
   }
 
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+  
+    // Get the input from the form and add it to json on /data
+    String quoteSubmission = getQuoteSubmission(request);
+    quotes.add(quoteSubmission);
+    // doGet seems to automatically make added submissions to json format. I'm not sure why.
```
Comment (jonathanjlin) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:54]: On line 40 in doGet, the quotes are converted to a json string

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -22,31 +22,41 @@
...
     response.getWriter().println(json);
     
   }
 
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+  
+    // Get the input from the form and add it to json on /data
+    String commentSubmission = request.getParameter("comment-submission");
+    comments.add(commentSubmission);
```
Comment (jonathanjlin) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:56]: If you get a post request before a get request request, comments will be null. You can mitigate this by initializing it in line 33: `private List<String> comments = new ArrayList<>()`, then you also won't need the null check in the get handler

Comment (BenLin00) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:72]: Implemented!

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import javax.servlet.http.HttpServletResponse;
+import com.google.gson.Gson;
+
+@WebServlet("/login")
+public class LoginServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("html/text");
+    String logInOutUrl;
```
[the remark this answers is not in the document]
Comment (BenLin00) [reply portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:35]: I Implemented logInOutUrl to be either a Login or a Logout url depending on the user status. 
logInOutUrl becomes a login url if the user is not yet logged in.
if the user is already logged in, logInOutUrl becomes a logout url.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -22,31 +22,75 @@
...
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    quotes = new ArrayList<>();
-    quotes.add("beep boop - ben lin");
-    quotes.add("They told me computers could only do arithmetic -Grace Hopper. I told them I cannot do arithmetic - Ben Lin");
-    quotes.add("hydrate or diedrate - Ben Lin");
-
-    // Convert the quotes to JSON using Gson library
-    String json = new Gson().toJson(quotes);
+    
+    UserService userService = UserServiceFactory.getUserService();
```
[the remark this answers is not in the document]
Comment (BenLin00) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:43]: Thanks for catching this!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -22,31 +22,75 @@
...
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
   
-  private ArrayList<String> quotes;
+  private ArrayList<String> comments = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (BenLin00) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:38]: Implemented!

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.HashMap;
+
+@WebServlet("/login")
+public class LoginServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("application/json");
+    String logInOutUrl;
+    int loginStatus;
```
Comment (jonathanjlin) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:36]: It's hard to tell here what this could contain, to make this clearer, one option is to make this 'boolean isLoggedIn', then when you need to serialize it to JSON, you can convert it to a 0 or 1. It looks like the javascript only needs a boolean, anyway, do you need to send an int?

File: portfolio/src/main/webapp/script.js [hunk 7]
```diff
@@ -27,24 +27,51 @@ function addRandomGreeting() {
...
+        // BUG: currently appends all comments each time "show comments" button is clicked
+    });
 }
 
-function getQuote() {
-    fetch('/data').then(response => response.json()).then((quotes) => {
-        // where the quote will go
-        quotesListElem = document.getElementById('quote-container');
+// redirect page to login upon "login/out" button click
+function loginButton() {
```
Comment (jonathanjlin) [inline portfolio/src/main/webapp/script.js:51]: Try to name functions as verbs, especially for callbacks. It's clearer to readers what onclick=login() or onclick=onLoginButtonClicked does rather than onclick=loginButton (login button what? hide it or log the user in or something else?)

File: portfolio/src/main/webapp/script.js [hunk 8]
```diff
@@ -27,24 +27,51 @@ function addRandomGreeting() {
...
+function hideLoginButton() {
+    var loginButton = document.getElementById("login-button");
+    var logoutButton = document.getElementById("logout-button");
 
-        // set the quote onto the page
-        quotesListElem.innerHTML = quote;
+    // fetch and hide/show a button. none == hide
+    fetch('/login').then(response => response.json()).then(data => {
+        if (data.loginStatus == 0) {
+            loginButton.style.display = "inline";
```
Comment (jonathanjlin) [inline portfolio/src/main/webapp/script.js:67]: Would it be better to set the button's class here and control how it's displayed via CSS?

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+    Gson gsonBuilder = new GsonBuilder().create();
+    HashMap loginDataMap = new HashMap();
```
Comment (jonathanjlin) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:39]: Try to avoid raw types: 'Map<String, Object> loginDataMap = new HashMap<>()'

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+@WebServlet("/login")
+public class LoginServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("application/json");
+    String logInOutUrl;
+    int loginStatus;
+
+    Gson gsonBuilder = new GsonBuilder().create();
```
Comment (jonathanjlin) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:38]: In general, declare variables as late as possible. If you define this at line 54, right before it's used, readers won't have to remember that this is in scope and whether it's being used before that line

Comment (BenLin00) [reply portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:38]: implemented & noted for future reference

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 9 above)
Comment (BenLin00) [reply portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:39]: What should I be using here instead to build a json map?

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 6 above)
Comment (BenLin00) [reply portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:36]: Nope! I had a brain fart and forgot the boolean data type existed. I changed it to boolean.

File: portfolio/src/main/webapp/script.js (hunk 8 above)
Comment (BenLin00) [reply portfolio/src/main/webapp/script.js:72]: resolved with 
loginButton.hidden = true;

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 11]
```diff
@@ -22,31 +22,72 @@
...
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.api.datastore.FetchOptions;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
   
-  private ArrayList<String> quotes;
+  private List<String> comments = new ArrayList<>();
```
Comment (jonathanjlin) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:39]: This can be final

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 12]
```diff
@@ -22,31 +22,72 @@
...
-
+    
     // Send the JSON as the response
     response.setContentType("application/json;");
-    response.getWriter().println(json);
     
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query = new Query("Message").addSort("timestamp", SortDirection.DESCENDING);
+    List<Entity> results = datastore.prepare(query).asList(FetchOptions.Builder.withLimit(10));
+    // add a limit
```
Comment (jonathanjlin) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:50]: nit: implementation comments that explain what code's doing usually goes either above the code it's explaining or to the end:

// add a limit
... FetchOptions.Builder.withLimit(10)
or
... FetchOptions.Builder.withLimit(10) // add a limit

though also try to limit the amount of these comments. In many cases they're redundant (such as this one, it's pretty clear that the query is using a limit), or it's a sign that a block of code can be extracted into a method for clarity. Consider:

// Query from Datastore 
DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
Query query = new Query("Message").addSort("timestamp", SortDirection.DESCENDING);
List<Entity> results = datastore.prepare(query).asList(FetchOptions.Builder.withLimit(10));

vs

List<Entity> results = queryFromDatastore();

private List<Entity> queryFromDatastore() {
  DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
  Query query = new Query("Message").addSort("timestamp", SortDirection.DESCENDING);
  return datastore.prepare(query).asList(FetchOptions.Builder.withLimit(10));
}

The second one helps keep methods focused and self-explanatory, as well as makes code more reusable, as any other method in this class can also call queryFromDatastore() without copying and pasting its logic

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 13]
```diff
@@ -22,31 +22,72 @@
...
+
+    String json = new Gson().toJson(commentsPopulate);
+    response.getWriter().println(json);
+
   }
 
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+  
+    // Get the input from the form and add it to json on /data
```
Comment (jonathanjlin) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:68]: Continuing on comments: here isn't the best place to say 'add it to json', since it's not being converted to JSON. Arguably the other 2 parts aren't needed either since 'request.getParameter' should be clear, and this class is a handler for '/data'. It's up to you, though, feel free to keep those if you think it helps remind readers

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 14]
```diff
@@ -22,31 +22,72 @@
...
+    String commentSubmission = request.getParameter("comment-submission");
+    comments.add(commentSubmission);
 
-}
+    UserService userService = UserServiceFactory.getUserService();
 
+    // Only logged-in users can post messages
```
Comment (jonathanjlin) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:74]: This is a good implementation level comment, it explains why there's a check for isUserLoggedIn

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+    response.setContentType("application/json");
+    // logInOutUrl is either the login or logout url depending on user status
+    String logInOutUrl;
+    boolean isLoggedIn;
+
+    // generate login/logout url & login status as 0/1
+    UserService userService = UserServiceFactory.getUserService();
+    if (userService.isUserLoggedIn()) {
+      logInOutUrl = userService.createLogoutURL("/");
+    //   ;StartPollingForCompletion();return false;">
```
Comment (jonathanjlin) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:43]: This doesn't look relevant to this code

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+public class LoginServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("application/json");
+    // logInOutUrl is either the login or logout url depending on user status
+    String logInOutUrl;
+    boolean isLoggedIn;
+
+    // generate login/logout url & login status as 0/1
```
Comment (jonathanjlin) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:39]: I think this comment is out of date now

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+      isLoggedIn = true;
+    } else {
+      logInOutUrl = userService.createLoginURL("/");
+      isLoggedIn = false;
+    }
+
+    Gson gsonBuilder = new GsonBuilder().create();
+    HashMap loginDataMap = new HashMap();
+
+    // add status and url to hashmap
```
Comment (jonathanjlin) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:53]: This comment isn't necessary, as it's clear that it's adding the URL and whether the user's logged in (rather than status) to the JSON. If it's doing something unusual or unintuitive, then here would be a good reason to explain why

File: portfolio/src/main/webapp/index.html [hunk 18]
```diff
@@ -7,38 +7,63 @@
     <link rel="stylesheet" href="style.css">
     <script src="script.js"></script>
   </head>
-
-  <body>
+<!--  -->
```
Comment (jonathanjlin) [inline portfolio/src/main/webapp/index.html:10]: Is this needed?

File: portfolio/src/main/webapp/script.js [hunk 19]
```diff
@@ -27,24 +27,52 @@ function addRandomGreeting() {
...
+        // where comments go
+        document.getElementById('comments-container').appendChild(ul);
+
+        // iterate through json object and make into html list
+        for (var key in commentsJson) {
+            let li = document.createElement('li');
+            li.innerHTML = commentsJson[key];
+            ul.appendChild(li);
+        }
+        // BUG: currently appends all comments each time "show comments" button is clicked
```
Comment (jonathanjlin) [inline portfolio/src/main/webapp/script.js:46]: I think you'll have to clear the comments-container's children at the beginning of this function

File: portfolio/src/main/webapp/script.js [hunk 20]
```diff
@@ -27,24 +27,52 @@ function addRandomGreeting() {
...
-        quotesListElem = document.getElementById('quote-container');
+// redirect page to login upon "login/out" button click
+function loginUser() {
+    fetch('/login').then(response => response.json()).then(data => {
+        window.location.replace(data.logInOutUrl);
+    });
 
-        //pick random quote
-        const quote = quotes[Math.floor(Math.random() * quotes.length)];
+    hideLoginButton(); // this function call is redundant because line 53 triggers hideLoginButton() onload
```
Comment (jonathanjlin) [inline portfolio/src/main/webapp/script.js:56]: If this is redundant, this comment should explain why, otherwise readers will wonder why it's not removed

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+      logInOutUrl = userService.createLogoutURL("/");
+    //   ;StartPollingForCompletion();return false;">
+      isLoggedIn = true;
+    } else {
+      logInOutUrl = userService.createLoginURL("/");
+      isLoggedIn = false;
+    }
+
+    Gson gsonBuilder = new GsonBuilder().create();
+    HashMap loginDataMap = new HashMap();
```
Comment (jonathanjlin) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:51]: I can't seem to add a new comment to the old comment I had about this... because of type erasure in Java, HashMap, HashMap, and HashMap<Object, Object> are roughly equivalent. Try to avoid the first two, though, since adding types helps static analysis tools. In this case I think since the keys are all Strings, you can declare this as a HashMap<String, Object>

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 9 above)
Comment (BenLin00) [reply portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java]: fixed.

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 21 above)
Comment (BenLin00) [reply portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:51]: Got it. Thanks.

File: portfolio/src/main/webapp/index.html (hunk 18 above)
Comment (BenLin00) [reply portfolio/src/main/webapp/index.html:10]: My mistake!

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 17 above)
Comment (BenLin00) [reply portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:53]: Thanks

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 11 above)
Comment (BenLin00) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:39]: changed.

File: portfolio/src/main/webapp/script.js (hunk 20 above)
Comment (BenLin00) [reply portfolio/src/main/webapp/script.js:56]: i removed it.

File: portfolio/src/main/webapp/script.js (hunk 19 above)
Comment (BenLin00) [reply portfolio/src/main/webapp/script.js:46]: Thanks

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 15 above)
Comment (BenLin00) [reply portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:43]: whoops! Reference code for patching a bug I forgot to delete.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 14 above)
Comment (BenLin00) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:74]: Thanks.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 13 above)
Comment (BenLin00) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:68]: Thanks

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 12 above)
Comment (BenLin00) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:50]: Great point. Thanks.
